email_pill: Add logic to convert email addresses with names attached into invite users email input to pills.

shuklamaneesh23 opened the pull request:
Fixes: #32874.

<!-- Describe your pull request here.-->

Fixes: <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

BartMassey commented:
Thanks for submitting this.

This looks like it allows full names in emails, but does not pass them on to be set as the user's full name in the subscription? If so, that would be less good, as the full name is useful information for the user's profile. I have used this in the past along with restricting users from changing full name to ensure that I know who I'm talking to in my course Zulips.

shuklamaneesh23 commented:
> Thanks for submitting this.
> 
> This looks like it allows full names in emails, but does not pass them on to be set as the user's full name in the subscription? If so, that would be less good, as the full name is useful information for the user's profile. I have used this in the past along with restricting users from changing full name to ensure that I know who I'm talking to in my course Zulips.

Ohh, I will look into the codebase for implementing  that as it might involve some backend changes to prepopulate the full name along with email. I have implemented this much as it is pretty simple frontend fix.

timabbott commented on web/src/email_pill.ts:
```diff
@@ -9,11 +9,17 @@ type EmailPill = {
+    const match = email_extraction_regex.exec(email);
+    if (match?.[1] !== undefined) {
+        email = match[1];
+    }
```
I'd prefer a function be responsible for this, and that it have tests. Also, how did you write the regex?

shuklamaneesh23 replied:
I have refactored the code to include a separate function for this and have also added tests for it.  

As for the regex used for email extraction, I asked `ChatGPT` to generate a regex that extracts the email from a string in the format `John <john@zulip.com>`. I then manually tested it by passing several values to ensure its correctness.

shuklamaneesh23 commented:
Hi @timabbott , I have updated the PR to use a separate function for that regex checking and added tests for that .
Along with that , I modified the backend part a little to repopulate the `full_name` field with disabled attribute such that the user can't change it as asked by @BartMassey in the above comment.

In order to achieve this i need to change the `Set[str]` to `Set[tuple[str,str]`.
Do let me know  if this approach best suit the current requirements  or I need to rework over it without changing the data structure used for holding the invitee email details.

Thank you, Looking forward to hear from you!

shuklamaneesh23 commented:
Hi @timabbott , would you like to take a look at it?
Thanks!

shuklamaneesh23 commented:
Hi @alya, could you please review this?
Thanks!

alya commented:
@sahil839 could you please review? I haven't tested.

Would be good to get this bug fixed for 10.0.

sahil839 commented on templates/zerver/register.html:
```diff
@@ -114,7 +114,7 @@ <h1>{{ _('Create your account') }}</h1>
                         <p class="fakecontrol">{{ full_name }}</p>
                     {% else %}
-                        <input id="id_full_name" class="required" type="text" name="full_name"
+                        <input id="id_full_name" class="required" type="text" name="full_name" {% if form.full_name.value() %}disabled="disabled"{% endif %}
```
And if we do not want to allow invited users to change their name while signing up then we should have code in backend too which does that.

shuklamaneesh23 replied:
@sahil839, I had a quick question. Currently, we don't have an option to allow or prevent users from changing their names in the invite modal.  

For this case, I’m considering the following: if someone invites a user with `John<john@zulip.com>`, we don’t want that person to change their username. However, if the email is not provided in this format, the user will be allowed to change their name.

Please let me know of exact behaviour what we want here.
Thanks!

sahil839 replied:
I am not sure. The issue does not mention anything, so would be good to discuss on CZO first.

sahil839 commented on zerver/actions/invites.py:
```diff
@@ -184,7 +184,7 @@ def check_invite_limit(realm: Realm, num_invitees: int) -> None:
 def do_invite_users(
     user_profile: UserProfile,
-    invitee_emails: Collection[str],
+    invitee_emails: Collection[tuple[str, str]],
```
Can this and similar variables below be renamed to make it clear that they are tuple containing email and name?

sahil839 replied:
Or maybe `invitee_names_and_emails`?

sahil839 commented on zerver/tests/test_invite.py:
```diff
@@ -2679,30 +2679,44 @@ class InviteeEmailsParserTests(ZulipTestCase):
+            ("email1@zulip.com", "Email One"),
+            ("email2@zulip.com", "EmailTwo"),
+            ("email3@zulip.com", "Email 3"),
+        }
```
There should be test which checks that when user signs up the name of the user is set to the one mentioned while inviting.

shuklamaneesh23 replied:
Is there a need to adding more tests as now we are not disabling the full_name field in the registration form.

timabbott commented on zerver/actions/invites.py:
```diff
@@ -270,10 +270,11 @@ def do_invite_users(
         # The logged in user is the referrer.
         prereg_user = PreregistrationUser(
             email=email,
+            full_name=full_name,
```
Let's add a comment here:

> We include a name if specified in the invitation request, but leave `full_name_validated=False`, so the receiving user can spell their name as they like after accepting the invitation.

timabbott commented on zerver/actions/invites.py:
```diff
@@ -238,20 +238,20 @@ def do_invite_users(
     gonna conflict with existing users
     """
-    error_dict = get_existing_user_errors(realm, good_emails)
+    error_dict = get_existing_user_errors(realm, {email for email, _ in good_emails})
```
`_` is not a good name for such a variable in a Django project; we usually use `ignored` here.

shuklamaneesh23 commented on zerver/actions/invites.py:
```diff
@@ -224,36 +225,36 @@ def do_invite_users(
             )
 
-    good_emails: set[str] = set()
+    good_invitees: set[Invitee] = set()
```
Renamed `good_emails` to `good_invitees` and `validated_emails` to `validated_invitees`.
Just let me know if there is some better variable name that we want to adopt here.

shuklamaneesh23 commented:
Hey @sahil839, I have updated the PR as per the discussion over the [CZO](https://chat.zulip.org/#narrow/channel/137-feedback/topic/User.20names.20in.20email.20invites).
Please have a look at it.
Thanks!

sahil839 commented on corporate/views/support.py:
```diff
@@ -584,16 +584,17 @@ def support(
     if query:
         key_words = get_invitee_emails_set(query)
 
+        key_word_emails = [key_word.email for key_word in key_words]
```
Thinking about this again, I am not sure how to name the variables here. I guess the query can be either email or full name but am not sure about the support subsystem in general, so we maybe we can have the variables on the basis of that.

Lets wait for feedback from Tim before makinga any changes here.

shuklamaneesh23 replied:
Hi @timabbott ,could you please tell me that what format we support in support queries so that I could address [this](https://github.com/zulip/zulip/pull/32881/files#r1938913536) comment?

sahil839 commented on web/src/email_pill.ts:
```diff
@@ -9,11 +9,25 @@ type EmailPill = {
+    return match?.[1] ?? null;
+}
 
 export function create_item_from_email(
```
Here also maybe we can rename the functions here to not just mention email. Email is needed to check already present emails, otherwise the display value of pill would be the whole text.

But not sure it is worth changing.

shuklamaneesh23 replied:
Here, I think we can let it be as email as it is still creating item from the email, however, its format has been changed to RFC 5322 format.

Or we can go with `Invitee` ?
Waiting for @timabbott confirmation.

sahil839 reviewed (commented):
I tested and it works good. I am not so sure about the naming changes, so you can wait for Tim's feedback on them.

shuklamaneesh23 commented:
Hi @sahil839 , this PR is ready for review other than the variable name changes.

Here, is a list of proposed name changes. If you'd like it , I would proceed with this.

- `invitee_emails_raw` => `invitees_raw`
- `get_invitee_emails_set` => `get_invitees_raw`
-  `invitee_emails_list` => `invitees_list`
- `is_email_with_name` => `is_invitee_with_name`
- `email` => `invitee`
- `create_item_from_email` => `create_item_from_invitee`
- `EmailPill` => `InvitePill`
-  `EmailPillWidget` => `InvitePillWidget`
- `get_email_from_item` => `get_invitee_from_item`
- `get_current_email` => `get_current_invitee`

sahil839 reviewed (commented):
Posted a couple of comments about backend changes. Do not have time to review the frontend changes right now. Will see if I can review them in next couple of days.

shuklamaneesh23 commented on zerver/views/invite.py:
```diff
@@ -181,14 +183,32 @@ def invite_users_backend(
-        invitee_emails.add(email.strip())
+            email = is_email_with_name.group("email").strip()
+        invitee_emails.add(Invitee(full_name="", email=email))
+
```
No longer required as per the discussion over [here](https://chat.zulip.org/#narrow/channel/137-feedback/topic/User.20names.20in.20email.20invites/near/2106774).

sahil839 replied:
I still think it is better to parse this in views code itself. And support code which uses this function can be updated accordingly. @timabbott thoughts?

shuklamaneesh23 replied:
@sahil839 , I've pushed an updated version of the code by putting the parsing logic inside the `invite_users_backend`.
Please have a look at that.
Is this what you were referrring to or still i'm missing it in some way.
Thanks!

sahil839 replied:
I was thinking to do similar to what we do for group settings, where the views code just accepts object like here it should accept object with name and email fields.

sahil839 commented on web/src/email_pill.ts:
```diff
@@ -1,43 +1,66 @@
     return {
-        type: "email",
-        email,
+        email: item.email,
```
Would be good to keep the `type` field here to follow the convention we have for other pills. But I am not sure what should be the value since we also have name along with email.

shuklamaneesh23 replied:
There was no `type` field previously either.  
I think Git was confused because it was tracking the changes in the wrong order.  
Simultaneous changes in two functions got messed up by Git.  
Now, I have fixed it.

sahil839 commented on web/src/input_pill.ts:
```diff
@@ -478,7 +479,10 @@ export function create<ItemType extends {type: string}>(
-            e.originalEvent.clipboardData?.setData("text/plain", store.get_text_from_item(item));
+            e.originalEvent.clipboardData?.setData(
+                "text/plain",
+                store.get_display_value_from_item(item),
```
I guess this will change the behavior for other pills, which I am not sure we want.

sahil839 commented on web/src/input_pill.ts:
```diff
@@ -28,7 +29,7 @@ type InputPillCreateOptions<ItemType> = {
         pill_config?: InputPillConfig,
     ) => ItemType | undefined;
-    get_text_from_item: (item: ItemType) => string;
+    get_text_from_item: (item: ItemType) => string | Invitees;
```
`get_text_from_item` is only used for copying, so this should just return a string always I guess. I think it can just return a string including both email and name or something like that.

shuklamaneesh23 commented:
Hi @sahil839 , here are certain updates from the discussion over the CZO threads:

- [api design](https://chat.zulip.org/#narrow/channel/378-api-design) > [Pre-populated full_name in the PreregistrationUser.](https://chat.zulip.org/#narrow/channel/378-api-design/topic/Pre-populated.20full_name.20in.20the.20PreregistrationUser.2E)

- [development help](https://chat.zulip.org/#narrow/channel/49-development-help) > [Backward compatibility](https://chat.zulip.org/#narrow/channel/49-development-help/topic/Backward.20compatibility)

**Updates**:

- frontend is now sending the emails in the str format rather than `stringified JSON object with full_name and email properties`
- A separate commit (commit 3) is added to allow comma into the name part of the email.
For ex:
```
 "John, Wick" <john@zulip.com>
```
- `get_invitee_emails_set` is as it is for `corporate.views.support`.
- `get_invitees_set` is defined in `zerver.views.invite.py` for parsing the string of RFC 5322 emails separated by commas or newlines.

sahil839 commented on web/src/email_pill.ts:
```diff
@@ -8,24 +10,32 @@ type EmailPill = {
+    const existing_emails = current_items
+        .map((item) => {
+            const parsed = parseOneAddress(item.email);
+            return parsed && parsed.type === "mailbox" ? parsed.address : null;
```
Rather than parsing this here as well, can we store a field like `parsed_email` in the item that can be used to compare here?

sahil839 commented on zerver/actions/invites.py:
```diff
@@ -230,36 +231,36 @@ def do_invite_users(
-            continue
+    for invitee in invitees:
+        if invitee.email == "":
+            continue  # nocoverage
```
Why does this not have coverage now? I don't think you have removed any tests.

shuklamaneesh23 replied:
The test `zerver.tests.test_invite.InviteUserTest.test_multi_user_invite` returns an empty email when passed to `get_invitee_emails_set`. However, when passed to `get_invitees_set`, it does not return an empty string.

Test:
```
 def test_multi_user_invite(self) -> None:
        """
        Invites multiple users with a variety of delimiters.
        """
        self.login("hamlet")
        # Intentionally use a weird string.
        self.assert_json_success(
            self.invite(
                """bob-test@zulip.com,     carol-test@zulip.com,
            dave-test@zulip.com


earl-test@zulip.com""",
                ["Denmark"],
            )
        )
        for user in ("bob", "carol", "dave", "earl"):
            self.assertTrue(find_key_by_email(f"{user}-test@zulip.com"))
        self.check_sent_emails(
            [
                "bob-test@zulip.com",
                "carol-test@zulip.com",
                "dave-test@zulip.com",
                "earl-test@zulip.com",
            ]
        )    
```

Output with get_invitee_emails_set:

```
{'', 'dave-test@zulip.com', 'carol-test@zulip.com', 'earl-test@zulip.com', 'bob-test@zulip.com'}
```

Output with get_invitees_set:

```
{Invitee(email='bob-test@zulip.com', full_name=''), Invitee(email='earl-test@zulip.com', full_name=''), Invitee(email='carol-test@zulip.com', full_name=''), Invitee(email='dave-test@zulip.com', full_name='')}
```

shuklamaneesh23 replied:
Yeah, we can remove it for now, but can also keep it as a defensive check for future changes in the codebase by adding a comment.

sahil839 commented on zerver/tests/test_invite.py:
```diff
@@ -2733,6 +2734,17 @@ def test_prereg_user_status(self) -> None:
         self.assertEqual(user.delivery_email, email)
 
+    def test_prereg_user_object_has_full_name_set(self) -> None:
+        email = "Test <test1@zulip.com>"
```
Would be good to test a couple of more formats like only email, or name including commas, etc.

sahil839 commented on web/src/email_pill.ts:
```diff
@@ -8,24 +10,32 @@ type EmailPill = {
-    if (!email_regex.test(email)) {
+    const original_email = email;
+    const parsed_address = parseOneAddress(email);
+    if (!parsed_address || parsed_address.type !== "mailbox") {
```
I am assuming you have checked that we do not want to support `type` values other than "mailbox".

shuklamaneesh23 replied:
According to current implementation where we are handling one email address at a time, it seems fine to restrict the type to `mailbox` only which is regular email address. If we also want to inclue `group`, then we probably need to handle that case separately.
I'll ask it on CZO.

sahil839 commented on web/src/input_pill.ts:
```diff
@@ -262,8 +262,21 @@ export function create<ItemType extends {type: string}>(
+                            : "",
+                    );
+                } else {
+                    pills = [pills].flat();
```
I don't get why is this needed.

sahil839 commented on web/src/input_pill.ts:
```diff
@@ -262,8 +262,21 @@ export function create<ItemType extends {type: string}>(
+            if (typeof pills === "string" && store.split_text_on_comma) {
                 pills = pills.split(/,/g).map((pill) => pill.trim());
+            } else {
+                const parsed_emails = parseAddressList(typeof pills === "string" ? pills : "");
```
Can't we just call this if `typeof pills === "string"`, since it does not do anything if it is a list of strings.

sahil839 commented on web/src/input_pill.ts:
```diff
@@ -262,8 +262,21 @@ export function create<ItemType extends {type: string}>(
+            if (typeof pills === "string" && store.split_text_on_comma) {
                 pills = pills.split(/,/g).map((pill) => pill.trim());
+            } else {
+                const parsed_emails = parseAddressList(typeof pills === "string" ? pills : "");
```
I think we should have this in a separate function which should be passed in `input_pill.create` and then just be called here when `split_text_on_comma` isn't true.

shuklamaneesh23 replied:
Hey @sahil839, please have another look.
Created a separate function `separate_emails_containing_comma_in_display_name` for this.
Please let me know if you are thinking of a better function name.
Thanks!

sahil839 commented on web/src/input_pill.ts:
```diff
@@ -368,7 +375,7 @@ export function create<ItemType extends {type: string}>(
             // Typing of the comma is prevented if the last field doesn't validate,
             // as well as when the new pill is created.
-            if (e.key === ",") {
+            if (e.key === "," && store.split_text_on_comma) {
```
Search pills also have `split_text_on_comma` set as `false`, so we should make sure that does not break the expected behavior for search pills. I am not completely sure what is the expected behavior with setting it to false for search pills, but maybe looking at the commit or PR which added that might give some idea.

shuklamaneesh23 replied:
@sahil839, This section of code is used to display the list again and to prevent typing a comma the first time in the search bar if a pill is already present. Below is the expected behavior.  

https://github.com/user-attachments/assets/dbed8daa-2ab1-4830-9ab8-d50ee5b45ee2


We can achieve the same behavior by replacing `split_text_on_comma` with `split_text_to_form_pills`, as the latter is currently defined only for email pills.  

I'm not sure if this is the best way to fix the behavior, but I can't think of another workaround.  
What do you think about this?

sahil839 replied:
I don't think there was any mistake in the previous code, just wanted to confirm about the behavior of search bar.

Regarding the code, I think logically using `split_text_on_comma` makes sense, and I think that should be used.

For search bar we allow typing "," and I don't think we want to create pill on entering "," in search bar, but probably @evykassirer can confirm once. Though I am not sure why this `keydown` handler is not being executed when typing letters and characters like "," but is executed when pressing keys like "Backspace" and "Enter".

evykassirer replied:
We don't currently create pills on "," -- but that doesn't mean we shouldn't! I think there are pros and cons to it, and if you are interested in changing the behavior it's worth opening a design thread on CZO.

evykassirer replied:
> Though I am not sure why this keydown handler is not being executed when typing letters and characters like "," but is executed when pressing keys like "Backspace" and "Enter".

I looked at the events on the search input:

<img width="483" alt="image" src="https://github.com/user-attachments/assets/ac0a0666-fdf2-4c51-8785-bfbd2f66264d" />

And eventually found that `maybeStopAdvance` calls `e.stopPropagation();` when typing letters and characters like `,`.

shuklamaneesh23 replied:
> We don't currently create pills on "," -- but that doesn't mean we shouldn't! I think there are pros and cons to it, and if you are interested in changing the behavior it's worth opening a design thread on CZO.

@evykassirer , In this PR, i'm not changing the behaviour of `search bar` in any way.(you can verify the behaviour on main branch and here).

All I'm changing here is the behaviour of `pill_container` of `invitee_emails_container` where previously, the text gets converted to pill when we press the comma key and know since the email type has been changed to `RFC5322` we want that comma to be typed.

| Before | After |
|---------|------|
| ![Before](https://github.com/user-attachments/assets/718edd1e-ac5f-4306-8fc7-450f114a6d03) |  ![After](https://github.com/user-attachments/assets/c8b9d796-710b-4b07-b070-75e14b5562ba) |

evykassirer replied:
Can you clarify what your question for me was?

sahil839 replied:
Yes, calling `e.stopPropagation` here seems the reason for `keydown` event not being called. My main concern here is the change in behavior of search bar due to adding `store.split_text_on_comma` condition as above, which seems correct logically.

Without this change, if there is already a pill in search bar and the user typed `,`, `,` was not inserted in the input and the typeahead was shown. 
But now after adding the `store.split_text_on_comma` condition, `,` is typed in the input and typeahead is shown accordingly.

shuklamaneesh23 replied:
This is now as per the discussion on the [CZO](https://chat.zulip.org/#narrow/channel/6-frontend/topic/How.20can.20i.20allow.20comma.20to.20be.20typed.20inside.20input.20pill-contain.2E/near/2163349).

sahil839 reviewed (commented):
Posted one comment. Looks good otherwise.

sahil839 commented on web/src/input_pill.ts:
```diff
@@ -278,6 +287,11 @@ export function create<ItemType extends {type: string}>(
+            // If nothing was successfully converted to pills, paste the original copied_text with shake effect.
+            if (copied_text && drafts.length === pills.length) {
+                store.$input.text(copied_text);
+                store.$input.addClass("shake");
```
This behavior was not present before to add shake animation. Was this discussed before?

shuklamaneesh23 replied:
The behaviour was already present on main when there is a wrong email in the string that we paste.


https://github.com/user-attachments/assets/be707cd0-8946-46e7-9cdc-351061c879ab

sahil839 replied:
Ok. I thought it was not present before as you added the specific code for it. Can it be like the `split_text_to_form_pills` returns the complete string if there is an invalid email so that `appendPill` can handle adding the shake animation?

shuklamaneesh23 replied:
OK, I've updated `split_text_to_form_pills` to return the complete string if there is an invalid email.

sahil839 commented on web/src/email_pill.ts:
```diff
@@ -49,6 +49,17 @@ export function get_current_email(
 }
 
+export function split_text_to_form_email_pills(raw_emails: string): string[] {
+    const parsed_emails = parseAddressList(raw_emails);
```
How does `parseAddressList` handle cases when only one email is valid and one isn't in comma seprated emails string? I see that none of them is converted to pill if I do that.

shuklamaneesh23 replied:
I tried using parseAddressList, but it couldn’t handle cases where some emails are valid and others are not — it fails to parse any of them. So, I’m using a different approach where I paste the entire string and apply a shake effect if there’s any invalid email.

sahil839 replied:
Ok, this changes the behavior which is present in main where valid email addresses are converted to pills. I am not sure whethere we want to try to maintain that behavior.

sahil839 replied:
Would be good to check if there is some config option that we can pass to `parseAddressList` for it to return the list of valid emails if string also contains invalid emails.

shuklamaneesh23 replied:
There is a  “partial” mode in this library that is `partial: true` flag (available when calling the parser with an options object). However, this does not cause parseAddressList to return a filtered list. Instead, it simply returns the AST of whatever it parsed up to the point of failure.

For example: 
```
String: John, john@gmail.com
Parsed: Error
```
 
```
String: john@gmail.com, john
Parsed: john@gmail.com
```

sahil839 commented on zerver/actions/invites.py:
```diff
@@ -230,36 +231,36 @@ def do_invite_users(
-        if email == "":
-            continue
+    for invitee in invitees:
+        if invitee.email == "":
```
I don't think this check is needed, as there is `if email` check in `get_invitees_set` and there is only one call to `do_invite_users` except tests.

sahil839 reviewed (commented):
Posted few comments.

shuklamaneesh23 commented on web/src/email_pill.ts:
```diff
@@ -49,6 +49,17 @@ export function get_current_email(
 
+export function split_text_to_form_email_pills(raw_emails: string): string[] {
+    const parsed_emails = parseAddressList(raw_emails);
+    if (!parsed_emails) {
```
Done @sahil839

alya commented:
From our live-reviews session:

1. Pasting a mix of valid and invalid inputs results in the whole thing not getting pillified. On CZO, valid inputs get pillified in this situation.
2. Deleting invalid inputs still doesn't cause the valid ones to get pillified. This issue might be obviated by fixing (1), but flagging as it might be indicative of a bug regardless.
3. When I just typed in names, they didn't get quotes around them, but somehow converting pasted names resulted in quotes. I don't think we want quotes at all: with pills, there's no need to further mark the names.

![Screenshot 2025-06-06 at 09 10 55](https://github.com/user-attachments/assets/2395b0a5-ff52-4d81-aca9-ac440d74e015)
4. With long names, can we abbreviate the name earlier and make sure we show the email? Showing as much of the email as we can should take priority, as that's the unique identifier here.

![Screenshot 2025-06-06 at 09 17 32](https://github.com/user-attachments/assets/b866e991-5bdd-4a5e-bb86-4c5a1c097371)

alya commented:
I haven't tested Enter-separated names, so that would be another thing to check.

alya commented:
[#design > duplicate invite emails feedback](https://chat.zulip.org/#narrow/channel/101-design/topic/duplicate.20invite.20emails.20feedback/with/2188597) for a potential UX improvement (not directly related to changes on this PR).

sahil839 commented on web/src/email_pill.ts:
```diff
@@ -1,43 +1,60 @@
+    const match = regex.exec(email);
+
+    if (match && (match[1]?.length ?? 0) > 20) {
+        email = `${match[1]?.slice(0, 20)}... ${match[2]}`;
```
Maybe we can have this logic in a separate function passed as `get_display_value_of_item`. And keep this function as before.

shuklamaneesh23 replied:
Done this.

sahil839 commented on web/src/input_pill.ts:
```diff
@@ -325,6 +332,12 @@ export function create<ItemType extends {type: string}>(
                 if (value.length > 0) {
+                    // If there are multiple values separated by commas, we should use insertManyPills
+                    // to handle them properly when pressing the Enter key.
+                    if (store.split_text_to_form_pills !== undefined && value.includes(",")) {
```
Is `if (!store.split_text_on_comma && value.includes(","))` a better condition here in terms of readability? Need to check if that breaks other cases where `split_text_on_comma` is `false`.

shuklamaneesh23 replied:
@sahil839 , it can be.
It is not breaking for other cases where `split_text_on_comma` is `false`.(only for email_pill and search_pill)

alya commented:
> This was already discussed in the [#feedback>User names in email invites@2172638](https://chat.zulip.org/#narrow/channel/137-feedback/topic/User.20names.20in.20email.20invites/near/2172638)

OK, posted a question on that thread.

alya commented:
I think @sahil839 can review/retest the rest.

sahil839 reviewed (commented):
Posted one comment and the truncation behavior is being discussed. Looks good other than that.

shuklamaneesh23 commented:
Hi @sahil839, I've added another commit as per discussion over the CZO to separate the `name` and `email` in the pill.
Please have a look at it and please let me know if it is required to squash it with the 1st one.
Thanks!


Update:

- `Name` will truncate untill email is long enough to be shown completely in the input field.

<img width="584" alt="Screenshot 2025-06-13 at 11 52 46 PM" src="https://github.com/user-attachments/assets/c5decf70-57bd-490f-929b-2dd840328d55" />

- `Email` will start truncating once it exceeds the input width like it was heppening previously.

<img width="586" alt="Screenshot 2025-06-13 at 11 53 21 PM" src="https://github.com/user-attachments/assets/e378bf55-8cce-4b99-9bea-8def6d0ef006" />

sahil839 commented on web/src/email_pill.ts:
```diff
@@ -63,6 +65,19 @@ export function split_text_to_form_email_pills(raw_emails: string): string[] {
+    const display_name = name?.replaceAll(/"([^"]*)"/g, "$1");
+
+    return render_input_pill({
+        is_email: true,
```
Would be good to name this as `is_email_pill`.

sahil839 commented on web/templates/input_pill.hbs:
```diff
@@ -20,8 +25,12 @@
             {{#if is_empty_string_topic}}
             {{sign}}topic: <span class="empty-topic-display">{{topic_display_name}}</span>
             {{else}}
+            {{#if is_email}}
```
This can use `else if` instead of using another if-else block inside `else` block.

sahil839 reviewed (commented):
Posted some comments.

sahil839 commented:
Looks good to me.

shuklamaneesh23 commented:
@sujalshah-bit , this PR is currently under integration review and I'm waiting for @timabbott to get this merge.

timabbott commented:
Closing in favor of https://github.com/zulip/zulip/pull/38508, which completes this work. Thanks @shuklamaneesh23!
